source/server/connection_handler_impl.cc

```diff
@@ -52,133 +56,142 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
-        }
-      } else {
-        auto v4_compatible_addr = address->ip()->ipv6()->v4CompatibleAddress();
-        // Remove this check when runtime flag
-        // `envoy.reloadable_features.strict_check_on_ipv4_compat` deprecated.
-        // If this isn't a valid Ipv4-mapped address, then do nothing.
-        if (v4_compatible_addr != nullptr) {
-          tcp_listener_map_by_address_.insert_or_assign(v4_compatible_addr->asStringView(),
-                                                        details);
+  for (auto& per_address_details : details->per_address_details_list_) {
+    // This map only store the new listener.
+    if (absl::holds_alternative<std::reference_wrapper<ActiveTcpListener>>(
+            per_address_details->typed_listener_)) {
+      tcp_listener_map_by_address_.insert_or_assign(per_address_details->address_->asStringView(),
+                                                    per_address_details);
+
+      auto& address = per_address_details->address_;
+      // If the address is Ipv6 and isn't v6only, parse out the ipv4 compatible address from the
+      // Ipv6 address and put an item to the map. Then this allows the `getBalancedHandlerByAddress`
+      // can match the Ipv4 request to Ipv4-mapped address also.
+      if (address->type() == Network::Address::Type::Ip &&
+          address->ip()->version() == Network::Address::IpVersion::v6 &&
+          !address->ip()->ipv6()->v6only()) {
+        if (address->ip()->isAnyAddress()) {
+          // Since both "::" with ipv4_compat and "0.0.0.0" can be supported.
+          // If there already one and it isn't shutdown for compatible addr,
+          // then won't insert a new one.
+          auto ipv4_any_address = Network::Address::Ipv4Instance(address->ip()->port()).asString();
+          auto ipv4_any_listener = tcp_listener_map_by_address_.find(ipv4_any_address);
+          if (ipv4_any_listener == tcp_listener_map_by_address_.end() ||
+              ipv4_any_listener->second->listener_->listener() == nullptr) {
+            tcp_listener_map_by_address_.insert_or_assign(ipv4_any_address, per_address_details);
+          }
+        } else {
+          auto v4_compatible_addr = address->ip()->ipv6()->v4CompatibleAddress();
+          // Remove this check when runtime flag
+          // `envoy.reloadable_features.strict_check_on_ipv4_compat` deprecated.
+          // If this isn't a valid Ipv4-mapped address, then do nothing.
+          if (v4_compatible_addr != nullptr) {
+            tcp_listener_map_by_address_.insert_or_assign(v4_compatible_addr->asStringView(),
+                                                          per_address_details);
+          }
         }
       }
+    } else if (absl::holds_alternative<std::reference_wrapper<Network::InternalListener>>(
+                   per_address_details->typed_listener_)) {
+      internal_listener_map_by_address_.insert_or_assign(
+          per_address_details->address_->asStringView(), per_address_details);
```

Comment:
Note that the use of `pre_address_details` and `per_address_details` is somewhat tricky and makes the code harder to read, so if one can use a completely different name for one of these it would be better.

Reply:
oops, `pre_address_details` should be a typo.